general/mem_manager.hpp

```diff
@@ -283,6 +283,19 @@ class Memory
        @note The current memory is NOT deleted by this method. */
    inline void Wrap(T *ptr, int size, MemoryType mt, bool own);
 
+   /** Wrap an externally pair of allocated pointers, @a h_ptr and @ d_ptr,
+       of the given host MemoryType @a h_mt. */
+   /** The new memory object will have the device MemoryType set as valid.
+
+       The given @a h_ptr and @a d_ptr must be allocated appropriately for the
+       given host MemoryType and its associated device MemoryType.
+
+       The parameter @a own determines whether both @a h_ptr and @a d_ptr will
+       be deleted when the method Delete() is called.
+
+       @note The current memory is NOT deleted by this method. */
+   inline void Wrap(T *h_ptr, T *d_ptr, int size, MemoryType h_mt, bool own);
```

Comment:
Nice!! I was wondering about ownership.

Reply:
I'm using the same `own` flag for both `host` and `device` pointer.
There are the `Memory`'s `OwnsDevicePtr`, `SetHostPtrOwner` and `SetDevicePtrOwner` that can be used to set the desired ownerships.

Reply:
I think the interface should be more flexible and ask for the types of both the host and the device memory separately. If the implementation does not support certain combinations, that will be an error, for now.

Also, it is good to document that `own` applies to both pointers and that users can call `SetXPtrOwner` if they need different ownerships for the host and the device pointers.

Reply:
Do we want to suggest by this _flexibility_ that the `mem_manager` will handle mixed memory types? One main assumption that is currently checked is that `h_mt` and `d_mt` are dual of each other.

The documentation already specifies the use of `own` for both pointers. I'll mention the `SetXPtrOwner` functions, good point!

Reply:
> Do we want to suggest by this _flexibility_ that the `mem_manager` will handle mixed memory types?

Yes, that's what I'm suggesting. Do you think this will be hard to support in the future?

By the way, this does not imply supporting all possible combinations, just that there are some "mixed" ones that may be useful.

Reply:
I don't think it would be _hard_, I'm more concerned to keep it simple.

This PR allows `Sundials` to work on GPU, maybe we can see if the work with `Hypre` will need more than this before adding mixed memory modes.

Reply:
Another reason I think it is good to add the device memory type to the signature: this makes it easier to understand the call -- there are two pointers but only one memory type which is a bit strange.

Anyway, if you don't want to change it, I'm not going to push you more. However, I think a description of the implied device memory type from the host memory type should be added to the doxygen comment.

Regarding sundials: I don't see how this PR is necessary to support sundials on gpus. Of course, I see how this PR may make sense in other contexts, I'm just curious about the connection to sundials you mentioned. As far as I know, sundials vectors do not allocate two (host+device) pointers. To me, it seems that we only need the ability to extract one of the pointers from the `Memory` class to wrap it as a sundials vector, and vice versa: given a sundials vector, extract its (single) pointer and wrap it as `mfem::Vector` with a given memory type (which then mfem can copy/move if necessary). In what case do you need to wrap two externally allocated pointers with sundials?

Reply:
I actually like the idea that host and device memory are dual from each other for now, it prevents bugs. If we always have to give both host and device `MemoryType`, I can easily see a lot of confusion in certain cases, with some very ugly calls to all sort of methods to access the `MemoryType`. I also think that if the need to do something more complex arises in the future, it wouldn't be hard, we're not closing any doors here. I also agree that the `MemoryManager` is already a complex class, and keeping it simple and clean is a priority since it's so important in the MFEM library.

Reply:
`{h_mt => d_mt}` description updated.